### ISSUE TITLE
Catched exception in a mocked method makes application crash

### ISSUE THREAD

**Issue by killkrt** [rollbear/trompeloeil#154]
## Description
I am mocking a method in order to raise an exception and I want to check if its caller _protects itself_ catching the exception correctly.

### Steps to reproduce
This is a very simplified version of my code:

```
struct Foo
{
    virtual void mayThrow() = 0;
    
    virtual bool throwerCaller() noexcpet
    {
        bool res = false;
        try
        {
            mayThrow();
            res = true;
        }
        catch(...)
        {
            res = false;
        }
        return res;
    }
};
```

The test should check whether the method `mayThrow` throws an exception, its caller method (`throwerCaller`) returns `false`.
I am using Doctest for writing my test case and using the proper Trompeloeil include for it.

```
    struct MockedFoo : Foo
    {
         MAKE_MOCK0( mayThrow(), void() );
    };

    TEST_CASE("Foo")
    {
         MockedFoo foo;
         foo.REQUIRE_CALL( foo, mayThrow() ).THROW( std::exception() );
         CHECK_FALSE( foo.throwerCaller() );
         // Even using CHECK_THROWS_AS( foo.throwCaller(), const std::exception& ) works
    }
```

Running this code it will abort due to a crash:
```
FATAL ERROR: test case CRASHED: SIGABRT - Abort (abnormal termination) signal
```

I've already checked with the main developer of Doctest (Viktor Kirilov) that this is issue is not related to Doctest

* doctest version: 2.3.3
* Trompeloeil version: commit 0ac46df 
* Operating System: macOS 10.14.6 and Arch Linux
* Compiler+version: clang 8.0.1

**Comment by rollbear** [comment]
There are a number of typos in the above (it doesn't compile as written) but when I rewrote the code as a simple single program calling the mock from `main()`, it works as expected. Can there be any significant info missing above? Here's what I tried:

```Cpp
#include <trompeloeil.hpp>

struct Foo
{
    virtual void mayThrow() = 0;

    virtual bool throwerCaller() noexcept
    {
        bool res = false;
        try
        {
            mayThrow();
            res = true;
        }
        catch(...)
        {
            res = false;
        }
        return res;
    }
};

struct MockedFoo : Foo
{
  MAKE_MOCK0( mayThrow, void() );
};

int main()
{
  MockedFoo f;
  REQUIRE_CALL(f, mayThrow()).THROW(std::exception());
  return f.throwerCaller();
}
```

It returns `0`, which is expected since the call throws, the exception is caught and the return value is `false`.

Can you provide a stack trace from the crash?

**Comment by killkrt** [comment]
Sorry @rollbear for the typos in my example code, I've just tried to simplify my code on the fly and I made some mistakes.

BTW, I've just realized that the crash was due to the fact that in the actual catch code I was writing a log message and the log message was using another pure virtual method that was not mocked.

The strange thing is that the debugger was reporting the crash at line where `mayThrow` is called instead on the catch part.

Anyway checking the full error log reports the clear explanation was there!

Shame on me!

Thank you for you support and sorry if I waste your time!
